src/gui/notifications/dbusnotifier.cpp

```diff
@@ -0,0 +1,224 @@
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ *
+ */
+
+#include "dbusnotifier.h"
+#include "notifications.h"
+#include <QDesktopServices>
+#include <QDir>
+#include <QUrl>
+#include <QWidget>
+#include <QtDBus/QDBusConnection>
+#include <QtCore/QDebug>
+
+namespace {
+    static const QLatin1String ACTION_NAME_DEFAULT("activate-context");
+    static const QLatin1String ACTION_NAME_OPEN_FINISHED_TORRENT("document-open"); // it must be named as the corresponding FDO icon
+    static const QLatin1String DBUS_SERVICE_NAME("org.freedesktop.Notifications");
+    static const QLatin1String DBUS_INTERFACE_NAME("org.freedesktop.Notifications");
+    static const QLatin1String DBUS_PATH("/org/freedesktop/Notifications");
+}
+
+DBusNotifier::DBusNotifier(QWidget *mainWindow, QObject *parent)
+    : Notifier(parent)
+    , mainWindow(mainWindow)
+    , notificationsService(new org::freedesktop::Notifications(DBUS_SERVICE_NAME, DBUS_PATH, QDBusConnection::sessionBus(), this))
+    , capabilities(getCapabilities(notificationsService))
+{
+    if (!QDBusConnection::sessionBus().isConnected()) {
+        qDebug() << "DBus is not connected";
+    }
+
+    bool connectionSuccess = QDBusConnection::sessionBus().connect(
+                 QString(), DBUS_PATH, DBUS_INTERFACE_NAME, QLatin1String("NotificationClosed"),
+                 this, SLOT(notificationClosed(uint,uint)));
+    assert(connectionSuccess);
+    if (!connectionSuccess) {
+        QString errMsg = QDBusConnection::sessionBus().lastError().message();
+        qDebug() << "Could not connect to DBus event. Error message from DBus is: " << errMsg;
+    }
+    connectionSuccess = QDBusConnection::sessionBus().connect(
+             QString(), DBUS_PATH, DBUS_INTERFACE_NAME, QLatin1String("ActionInvoked"),
+             this, SLOT(actionInvoked(uint,QString)));
+    assert(connectionSuccess);
+    if (!connectionSuccess) {
+        QString errMsg = QDBusConnection::sessionBus().lastError().message();
+        qDebug() << "Could not connect to DBus event. Error message from DBus is: " << errMsg;
+    }
+}
+
+DBusNotifier::~DBusNotifier()
+{
+    QDBusConnection::sessionBus().disconnect(
+                 QString(), DBUS_PATH, DBUS_INTERFACE_NAME, QLatin1String("ActionInvoked"),
+                 this, SLOT(actionInvoked(uint,QString)));
+    QDBusConnection::sessionBus().disconnect(
+                     QString(), DBUS_PATH, DBUS_INTERFACE_NAME, QLatin1String("NotificationClosed"),
+                     this, SLOT(notificationClosed(uint,uint)));
+}
+
+DBusNotifier::CapabilityFlags DBusNotifier::getCapabilities(org::freedesktop::Notifications *notificationsService)
+{
+    QDBusPendingReply<QStringList> reply = notificationsService->GetCapabilities();
+    reply.waitForFinished();
+    CapabilityFlags res(0);
+
+    QStringList capabilities = reply.value();
+    if(capabilities.contains(QLatin1String("action-icons"))) {
```

Comment:
Although using braces in such cases (single statement in block) is not forbidden, try to avoid this (if it does not matter for you).